### PR TITLE
usb: hid: allow boot interface Protocol Code to be set per device 

### LINKED ISSUE #32778: Cannot support both HID boot report keyboard and mouse on a USB HID device
**Describe the bug**
With current Zephyr implementation, the USB HID instances share a single HID boot report type (set as a Kconfig option). If we add support for defining the HID boot report type separately for every USB HID instance, we could send boot reports as both mouse and keyboard to the host computer in boot report mode.

**Expected behavior**
USB HID device with multiple HID instances should be able to forward HID boot reports both as mouse and keyboard.

**Impact**
Cannot design USB HID device that supports HID boot reports from both mouse and keyboard.

**Environment**
 - Commit SHA: 48321ef2525c14c4d9b6040afe8ae72ff07a8d1e

### PR DESCRIPTION
**Pull request by jfischer-no**
Kconfig option USB_HID_PROTOCOL_CODE does not allow to set
boot interface protocol code for specific HID device but
only to set the same value for all device.
Add new API function to allow the application to set
Protocol Code per device. Deprecate USB_HID_PROTOCOL_CODE option.

Fixes: #32778

### CODE REVIEW

File: subsys/usb/class/hid/core.c [hunk 1]
```diff
@@ -675,6 +675,19 @@ static void hid_interface_config(struct usb_desc_header *head,
 		.endpoint = hid_ep_data_##x,				\
 	};
 
+int usb_hid_set_proto_code(const struct device *dev, uint8_t proto_code)
+{
+	struct usb_cfg_data *cfg = (void *)dev->config;
+	struct usb_if_descriptor *if_desc = (void *)cfg->interface_descriptor;
+
+	if (IS_ENABLED(CONFIG_USB_HID_BOOT_PROTOCOL)) {
```
[the remark this answers is not in the document]
Comment (jfischer-no) [reply subsys/usb/class/hid/core.c:683]: I think no, reserved may be still valid.

File: subsys/usb/class/hid/Kconfig [hunk 2]
```diff
@@ -44,12 +44,18 @@ config USB_HID_POLL_INTERVAL_MS
 	  Polling interval in ms selected by the USB HID Device.
 
 config USB_HID_REPORTS
```
[the remark this answers is not in the document]
Comment (jfischer-no) [reply subsys/usb/class/hid/Kconfig:46]: I guess yes, but how I see it just user experience there is no benefit from this change in the code. And that needs to be done in #32355 :-)

File: subsys/usb/class/hid/core.c [hunk 3]
```diff
@@ -675,6 +675,19 @@ static void hid_interface_config(struct usb_desc_header *head,
 		.endpoint = hid_ep_data_##x,				\
 	};
 
+int usb_hid_set_proto_code(const struct device *dev, uint8_t proto_code)
+{
+	struct usb_cfg_data *cfg = (void *)dev->config;
+	struct usb_if_descriptor *if_desc = (void *)cfg->interface_descriptor;
```
Comment (MarekPieta) [inline subsys/usb/class/hid/core.c:681]: Do we need the explicit casts here ?

**Review by MarekPieta — APPROVED** [review]

File: subsys/usb/class/hid/core.c (hunk 3 above)
Comment (jfischer-no) [reply subsys/usb/class/hid/core.c:681]: ohh, copy & paste from above, but `void *interface_descriptor;` does not need to have const in struct usb_cfg_data().

Comment (jfischer-no) [reply subsys/usb/class/hid/core.c]: fixed it for this patch.

**Review by MarekPieta — APPROVED** [review]

**Review by anangl — APPROVED** [review]

**Pull request merged**